### PR TITLE
Action bars for window (un-)installation

### PR DESCRIPTION
**Pull request by Kubius**
[FEAT][REWORK]
## About the PR 

Introduces generic action bars to the use of tools on windows for installation or uninstallation, and shuffles around some code accordingly. Some notes:

- Generic action bars were put in place for each step. A calculation is done before each to take bonues from user traits into account.
- The deconstruct time that already existed has been uncommented. Windows without a deconstruct time will be instantly screwed and pried, bypassing the action bars entirely.
- Thin windows can still be rotated (tested ingame) and appear to behave appropriately.

## Why's this needed? 

More action bars generally good, and particularly benefits reinforced windows (harder to take them apart when someone's trying to stop you).

## Changelog

```changelog
(u)Kubius
(*)Window disassembly now involves action bars, with more reinforced windows requiring more disassembly time.
```

### CODE REVIEW

**Comment by Kubius** [comment]
Error on Kondaru due to random room checking goof. Not related to PR contents.

**Comment by Kubius** [comment]
Exact speed of window disassembly can be tuned quite easily with a change to the disassembly time; if it feels like it's the wrong interval for gameplay fluidity, the time can go up or down for different window types as needed. Right now, it's *fairly* prompt, but not what I'd call outright fast.

**Comment by ThatFiveGuys** [comment]
The thing is players dont measure what they do on spreadsheets, they measure it on how things feel to do. Even if the action bars are .1 seconds the existence of a countdown makes it feel way slower then no countdown.

**Comment by Kubius** [comment]
That's how action bars work, and I don't know that I agree on it being a net negative to have them at all for "lesser" windows. That said, these changes support **total bypass of the action bar**; if there's broad agreement that old disassembly is better for lesser windows, it's a simple matter of changing one defined value later on and the code will handle it accordingly.

**Comment by ZeWaka** [comment]
it's not a 'random room checking goof,' it's a "my PR branch is not up to date goof"

**Comment by Kubius** [comment]
[minor semantic haggling: it was a (random room) (checking goof), not a (random) (room checking goof), but yes it got fixed in master]

Standardized and lowered some timings in 0e5451a. Most windows now have a deconstruct_time of 10, basic crystal windows are at 20, and reinforced crystal are at 50. Here's a demo of windows with a deconstruct_time of 10 and no Carpenter trait or engineering training.

https://user-images.githubusercontent.com/12454065/177454321-b23b1dfc-e0f7-4908-babb-e7d9a7584ad7.mp4

File: code/obj/window.dm [hunk 1]
```diff
@@ -26,7 +26,7 @@
 	var/default_material = "glass"
 	var/default_reinforcement = null
 	var/reinf = 0 // cant figure out how to remove this without the map crying aaaaa - ISN
-	var/deconstruct_time = 0//20
+	var/deconstruct_time = 10
```
Comment (flrsh) [inline code/obj/window.dm:29]: 10 deciseconds? 1 second?

File: code/obj/window.dm [hunk 2]
```diff
@@ -371,67 +371,55 @@
...
-						return
-				src.anchored = !(src.anchored)
-				src.stops_space_move = !(src.stops_space_move)
-				user.show_text("You have [src.anchored ? "fastened the frame to" : "unfastened the frame from"] the floor.", "blue")
-				logTheThing("station", user, null, "[src.anchored ? " anchored" : " unanchored"] [src] at [log_loc(src)].")
-				return 1
+					SETUP_GENERIC_ACTIONBAR(user, src, total_decon_time, /obj/window/proc/assembly_handler, list(user,W), W.icon, W.icon_state,null,null)
+				else
+					state = 3 - state
+					user.show_text("You have [state == 1 ? "unfastened the window from" : "fastened the window to"] the frame.", "blue")
```
Comment (flrsh) [inline code/obj/window.dm:388]:
```suggestion
					assembly_handler(user, W)
```

File: code/obj/window.dm [hunk 3]
```diff
@@ -371,67 +371,55 @@
...
+						if (H.traitHolder.hasTrait("carpenter") || H.traitHolder.hasTrait("training_engineer"))
+							total_decon_time = round(total_decon_time / 2)
+					user.show_text("You begin to [src.anchored ? "unfasten the frame from" : "fasten the frame to"] the floor...", "red")
+					SETUP_GENERIC_ACTIONBAR(user, src, total_decon_time, /obj/window/proc/assembly_handler, list(user,W), W.icon, W.icon_state,null,null)
+				else
+					src.anchored = !(src.anchored)
+					src.stops_space_move = !(src.stops_space_move)
+					user.show_text("You have [src.anchored ? "fastened the frame to" : "unfastened the frame from"] the floor.", "blue")
+					logTheThing("station", user, null, "[src.anchored ? " anchored" : " unanchored"] [src] at [log_loc(src)].")
+					src.align_window()
```
Comment (flrsh) [inline code/obj/window.dm:404]:
```suggestion
					assembly_handler(user, W)
```

File: code/obj/window.dm [hunk 4]
```diff
@@ -371,67 +371,55 @@
...
+					if(ishuman(user))
+						var/mob/living/carbon/human/H = user
+						if (H.traitHolder.hasTrait("carpenter") || H.traitHolder.hasTrait("training_engineer"))
+							total_decon_time = round(total_decon_time / 2)
+					user.show_text("You begin to [src.state ? "pry the window out of" : "pry the window into"] the frame...", "red")
+					SETUP_GENERIC_ACTIONBAR(user, src, total_decon_time, /obj/window/proc/assembly_handler, list(user,W), W.icon, W.icon_state,null,null)
+				else
+					playsound(src.loc, "sound/items/Crowbar.ogg", 75, 1)
+					state = 1 - state
+					user.show_text("You have [src.state ? "pried the window into" : "pried the window out of"] the frame.", "blue")
```
Comment (flrsh) [inline code/obj/window.dm:422]:
```suggestion
					assembly_handler(user, W)
```

**Review by flrsh — CHANGES_REQUESTED** [review]
Generally a fan, I like being able to tell how long these kinds of actions take.

File: code/obj/window.dm (hunk 1 above)
Comment (Kubius) [reply code/obj/window.dm]: All spots that set deconstruct time now use SECOND/S for clarity.

**Comment by Kubius** [comment]
Checked ingame after moving to assembly_handler use in those spots, appears to have continued working as it's supposed to.

**Review by flrsh — APPROVED** [review]

**Pull request merged**